### DISCUSSION TITLE
kubernetes-maven-plugin automatic pull for cacheFrom image

### DISCUSSION THREAD

**Discussion by davidecavestro** [eclipse/jkube#1519] [General]
**Forewords**
Hi, I open a discussion here as per @manusa [suggestion](https://gitter.im/eclipse/jkube?at=628b90b209eea00adeab4366).
I've not provided some reproducible code yet as it would imply pushing a container image on a shared docker registry.

**The issue**
I've configured the _kubernetes-maven-plugin_ with the _cacheFrom_ config entry in order to reuse some layers from previously produced images.
But it seems that I have to force pulling the image externally in order to have the _cacheFrom image_ in-place and eventually reuse its layers.

**Additional details**
I leveraged multi-staged builds extracting layers produced by spring boot maven plugin into subfolders, so that when 3rd party deps don't change within the fatjar I reuse the old layer.
I've seen this can be achieved even on ephemeral builders (no docker cache) using _cacheFrom_ to pass the reference to a previously built image (i.e. the image built on the previous run by the CI tool).

I've tried passing to maven CLI _-Djkube.docker.imagePullPolicy=Always -Djkube.docker.autoPull=on_ but they seem to be inifluent, though I have no debugged so far.
OTOH I see a reuse of layers if I manually force the download of the _cacheFrom_ image from outside maven (or even with a maven exec).

**Excerpt from Dockerfile**
Follows an excerpt from Dockerfile, where the various `COPY --from=builder` instructions at the bottom potentially benefit from the _cacheFrom_ image (when locally pulled).
 
<pre>
# builder stage for fatjar layers extraction
FROM registry.access.redhat.com/ubi8/openjdk-17 as builder
USER root
RUN mkdir -p /inflated
WORKDIR /inflated
COPY maven/jars/app.jar app.jar
# explode the fatjar into a subfolder per layer
RUN java -Djarmode=layertools -jar app.jar extract

# restart from scratch
FROM registry.access.redhat.com/ubi8/openjdk-17

# add layers content
COPY --from=builder /inflated/dependencies/ ./
COPY --from=builder /inflated/spring-boot-loader/ ./
COPY --from=builder /inflated/snapshot-dependencies/ ./
COPY --from=builder /inflated/application/ ./

</pre>

**Comment by manusa** [comment]
Hi @davidecavestro,

This has nothing to do with your original question, but while we investigate:

I understand that the end purpose is to be able to reuse some of the intermediate images whilst performing your final build.

Have you tried doing a [multilayered](https://www.youtube.com/watch?v=0t-bLulywSM&t=33s) (XML/DSL configured) build with a non-fatjar?

In your case, considering the build produces an intermediate version of your jar with the dependencies, it would be something like:
```xml
<configuration>
  <images>
    <image>
      <name>${image.repo}/${image.name}:%l</name>
      <build>
        <from>registry.access.redhat.com/ubi8/openjdk-17</from>
        <assembly>
          <targetDir>/home/jboss</targetDir>
          <excludeFinalOutputArtifact>true</excludeFinalOutputArtifact>
          <layers>
            <layer>
              <id>dependencies</id>
              <fileSets>
                <fileSet>
                  <directory>${project.build.directory}/dependencies</directory>
                </fileSet>
              </fileSets>
            </layer>
            <layer>
              <id>loader</id>
              <fileSets>
                <fileSet>
                  <directory>${project.build.directory}/loader</directory>
                </fileSet>
              </fileSets>
            </layer>
            <!-- ... -->
            <layer>
              <id>app</id>
              <fileSets>
                <fileSet>
                  <directory>${project.build.directory}/app</directory>
                </fileSet>
              </fileSets>
            </layer>
          </layers>
        </assembly>
        <ports>
          <!-- Ports to expose in the pod specs -->
          <port>8080</port>
        </ports>
      </build>
    </image>
  </images>
</configuration>
```

**Comment by davidecavestro** [comment]
Sure thanks. Sorry, I've just realized that I never closed this discussion.
For the records, I am now using the _spring-boot_ generator along with _jib_ buildStrategy to get both multi-layered, multi-arch images.
